src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -146,90 +191,142 @@ private async ValueTask LoadOrReloadProjectsAsync(ImmutableSegmentedList<Project
         }
     }
 
+    /// <summary>
+    /// Loads a project in the MSBuild host.
+    /// </summary>
+    protected abstract Task<(RemoteProjectFile projectFile, bool hasAllInformation, BuildHostProcessKind preferred, BuildHostProcessKind actual)?> TryLoadRemoteProjectAsync(
+        BuildHostProcessManager buildHostProcessManager, string projectPath, CancellationToken cancellationToken);
+
     /// <returns>True if the project needs a NuGet restore, false otherwise.</returns>
     private async Task<bool> LoadOrReloadProjectAsync(ProjectToLoad projectToLoad, ToastErrorReporter toastErrorReporter, BuildHostProcessManager buildHostProcessManager, CancellationToken cancellationToken)
     {
         BuildHostProcessKind? preferredBuildHostKindThatWeDidNotGet = null;
         var projectPath = projectToLoad.Path;
+        Contract.ThrowIfFalse(PathUtilities.IsAbsolute(projectPath));
+
+        // Before doing any work, check if the project has already been unloaded.
+        using (await _gate.DisposableWaitAsync(cancellationToken))
+        {
+            if (!_loadedProjects.ContainsKey(projectPath))
+            {
+                return false;
+            }
+        }
 
         try
         {
-            var preferredBuildHostKind = GetKindForProject(projectPath);
-            var (buildHost, actualBuildHostKind) = await buildHostProcessManager.GetBuildHostWithFallbackAsync(preferredBuildHostKind, projectPath, cancellationToken);
+            if (await TryLoadRemoteProjectAsync(buildHostProcessManager, projectPath, cancellationToken)
+                is not var (remoteProjectFile, hasAllInformation, preferredBuildHostKind, actualBuildHostKind))
+            {
+                _logger.LogWarning($"Unable to load project '{projectPath}'.");
```

Comment:
will the reason it didn't load be logged anywhere?  just trying to figure out what a user does at this point.

Reply:
Here the user can file a bug.

Reply:
It looks like yes, the reason will be logged, LogDiagnosticsAsync is called inside here.